app/models/user.rb

```diff
@@ -28,4 +28,20 @@ def member_of?(organisation)
   def application_names
     memberships.map(&:application_names).flatten.uniq
   end
+
+  def roles_for_application(application_id)
+    memberships
+      .joins(:application_memberships)
+      .where(application_memberships: {application_id: application_id})
+      .map{|m| m.roles_for_application(application_id)}
+      .flatten.uniq
+  end
+
+  def self.current
+    Thread.current[:user]
+  end
+
+  def self.current=(user)
+    Thread.current[:user] = user
+  end
```

Comment:
@aldavidson as long as `User.current` is only ever accessed in a controller/request context then that should be okay - just bear it in mind if you see some weird errors.